split-llvm-extract.cpp

```diff
@@ -59,42 +62,82 @@ llvm::cl::opt<bool> dry("d",
+      // TODO: ignore null's for now.
+    } else if (auto globalVariable =
+                   llvm::dyn_cast<llvm::GlobalVariable>(currentOperand)) {
+      if (globalVariable->isConstant()) {
+        globals.insert(globalVariable);
+      } 
+    } else if (auto user = llvm::dyn_cast<llvm::User>(currentOperand)) {
+      for (uint32_t i = 0; i < user->getNumOperands(); i++) {
+        auto nextOperand = user->getOperand(i);
+        if (auto instruction =
+                llvm::dyn_cast<llvm::Instruction>(nextOperand)) {
+          continue;
+        } else {
+          operands.push(nextOperand);
+        }
+      }
+    }
+  }
+  return globals;
+}
 
-    for (uint32_t i = 0; i < instruction.getNumOperands(); i++) {
-      operands.push(instruction.getOperand(i));
+struct MyPass : public llvm::InstVisitor<MyPass> {
+  std::unordered_set<llvm::GlobalVariable *> globals;
+  // TODO: Move this to a separate function because globals 
+  // also need the full dependency tree moved.
+  void visitInstruction(llvm::Instruction &instruction) {
+    std::queue<llvm::GlobalVariable *> queueVariablesNeedMoving;
+    {
+      auto result = iterateOperands(instruction);
+      for(const auto item : result) {
+        queueVariablesNeedMoving.push(item);
+      }
     }
+    while(queueVariablesNeedMoving.size() > 0) {
+      auto currentVariable = queueVariablesNeedMoving.front();
+      queueVariablesNeedMoving.pop();
+      globals.insert(currentVariable);
 
-    while (operands.size() > 0) {
-      auto currentOperand = operands.top();
-      operands.pop();
-
-      if (currentOperand == nullptr) {
-        // TODO: ignore null's for now.
-      } else if (auto globalVariable =
-                     llvm::dyn_cast<llvm::GlobalVariable>(currentOperand)) {
-        if (globalVariable->isConstant()) {
-          globals.insert(globalVariable);
-        } 
-      } else if (auto user = llvm::dyn_cast<llvm::User>(currentOperand)) {
-        for (uint32_t i = 0; i < user->getNumOperands(); i++) {
-          auto nextOperand = user->getOperand(i);
-          if (auto instruction =
-                  llvm::dyn_cast<llvm::Instruction>(nextOperand)) {
-            continue;
-          } else {
-            operands.push(nextOperand);
+      if(currentVariable->hasInitializer()) {
+        auto alsoNeedMvoing = iterateOperands(*currentVariable->getInitializer());
+        for(const auto item : alsoNeedMvoing) {
+          if(globals.find(item) == globals.end()) {
+            queueVariablesNeedMoving.push(item);
           }
         }
-      }
+     }
     }
   }
 };
 
+std::string getFileName(llvm::GlobalObject &value){
+  llvm::SmallVector<std::pair<unsigned, llvm::MDNode *>, 4> MDs;
+  value.getAllMetadata(MDs);
+  for (auto &MD : MDs) {
+    if (llvm::MDNode *N = MD.second) {
+      if (auto *subProgram = llvm::dyn_cast<llvm::DISubprogram>(N)) {
+        return subProgram->getFilename().str();
+      }
+    }
+  }
+  return "";
```

Comment:
Should we `err` in this case?

Reply:
The llvm code base would have done the some which I know doesn't make it fine, but this is printing only for diagnostics, so it doesn't matter too much.

Reply:
OK, can we add a doc comment to the function saying something like "returns `""` if nothing is found"?

Reply:
fixed: https://github.com/capablevms/llvm-function-split/pull/9/commits/f879b54e102f230620f99ec6222ca1565d1d38fa